pkg/pushsync/pushsync.go

```diff
@@ -279,3 +296,81 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk) (rr *pb.R
 
 	return nil, topology.ErrNotFound
 }
+
+// closestPeer returns address of the peer that is closest to the chunk with
+// provided address addr. This function will ignore peers with addresses
+// provided in skipPeers
+func (ps *PushSync) closestPeer(addr swarm.Address, skipPeers ...swarm.Address) (swarm.Address, error) {
+	closest := swarm.Address{}
+	err := ps.peerSuggester.EachPeer(func(peer swarm.Address, po uint8) (bool, bool, error) {
+		for _, a := range skipPeers {
+			if a.Equal(peer) {
+				return false, false, nil
+			}
+		}
+		if closest.IsZero() {
+			closest = peer
+			return false, false, nil
+		}
+		dcmp, err := swarm.DistanceCmp(addr.Bytes(), peer.Bytes(), closest.Bytes())
+		if err != nil {
+			return false, false, fmt.Errorf("distance compare error. addr %s closest %s peer %s: %w", addr.String(), closest.String(), peer.String(), err)
+		}
+		if dcmp == 1 {
+			closest = peer
+		}
+		return false, false, nil
+	})
+
+	if err != nil {
+		return swarm.Address{}, err
+	}
+
+	if closest.IsZero() {
+		return swarm.Address{}, topology.ErrNotFound
+	}
+
+	dcmp, err := swarm.DistanceCmp(addr.Bytes(), closest.Bytes(), ps.address.Bytes())
+	if err != nil {
+		return swarm.Address{}, fmt.Errorf("distance compare addr %s closest %s base address %s: %w", addr.String(), closest.String(), ps.address.String(), err)
+	}
+
+	if dcmp == 0 {
+		return closest, topology.ErrWantSelf
+	}
+
+	return closest, nil
+}
+
+func (ps *PushSync) pushToPeers(ctx context.Context, ch swarm.Chunk) (err error) {
+
+	skip := []swarm.Address{ps.address}
```

Comment:
this is incorrect - this should be the origin address, not the base address. also, for uniformity, I would add a `origin swarm.Address` and `allowUpstream bool` flag to `pushToClosest` method, and include the origin address when doing the call to `pushToClosest` when forwarding (`allowUpstream=false` in this case). When calling it from `PushToClosest` then `allowUpstream=true`.

Reply:
I believe base address is correct here because we do not want the final storing node to send the chunk to itself again. Origin address should be able to receive the chunk if the storing node decides that it is the closest after itself.

 I need a bit more clarification on the other comments.

Reply:
A node cannot send to itself. You will never get the base address from the kademlia iterator, if you do - we have a serious issue. This should never happen.

> Origin address should be able to receive the chunk if the storing node decides that it is the closest after itself

That is indeed true

Reply:
ok